File: en-html-utf8/calendar.js
function JHashMap()
{
    this.length = 0;
    this.prefix = "hashmap_prefix_20040716_";
}

JHashMap.prototype.put = function (key, value)
{
    this[this.prefix + key] = value;
    this.length ++;
}

JHashMap.prototype.get = function(key)
{
    return typeof this[this.prefix + key] == "undefined" 
            ? null : this[this.prefix + key];
}

JHashMap.prototype.keySet = function()
{
    var arrKeySet = new Array();
    var index = 0;
    for(var strKey in this)
    {
        if(strKey.substring(0,this.prefix.length) == this.prefix)
            arrKeySet[index ++] = strKey.substring(this.prefix.length);
    }
    return arrKeySet.length == 0 ? null : arrKeySet;
}

JHashMap.prototype.values = function()
{
    var arrValues = new Array();
    var index = 0;
    for(var strKey in this)
    {
        if(strKey.substring(0,this.prefix.length) == this.prefix)
            arrValues[index ++] = this[strKey];
    }
    return arrValues.length == 0 ? null : arrValues;
}

JHashMap.prototype.size = function()
{
    return this.length;
}

JHashMap.prototype.remove = function(key)
{
    delete this[this.prefix + key];
    this.length --;
}

JHashMap.prototype.clear = function()
{
    for(var strKey in this)
    {
        if(strKey.substring(0,this.prefix.length) == this.prefix)
            delete this[strKey];   
    }
    this.length = 0;
}

JHashMap.prototype.isEmpty = function()
{
    return this.length == 0;
}

JHashMap.prototype.containsKey = function(key)
{
    for(var strKey in this)
    {
       if(strKey == this.prefix + key)
          return true;  
    }
    return false;
}

JHashMap.prototype.containsValue = function(value)
{
    for(var strKey in this)
    {
       if(this[strKey] == value)
          return true;  
    }
    return false;
}

JHashMap.prototype.putAll = function(map)
{
    if(map == null)
        return;
    if(map.constructor != JHashMap)
        return;
    var arrKey = map.keySet();
    var arrValue = map.values();
    for(var i in arrKey)
       this.put(arrKey[i],arrValue[i]);
}

//toString
JHashMap.prototype.toString = function()
{
    var str = "";
    for(var strKey in this)
    {
        if(strKey.substring(0,this.prefix.length) == this.prefix)
              str += strKey.substring(this.prefix.length) 
                  + " : " + this[strKey] + "\r\n";
    }
    return str;
}

function strncasecmp(src1, src2, len)
{
	if(src1.length < len || src2.length < len)
		return false;
		
	var tmp1 = src1.toUpperCase();
	var tmp2 = src2.toUpperCase();
	
	var tmp3 = tmp1.substr(0, len);
	var tmp4 = tmp2.substr(0, len);
	
	if (tmp3 == tmp4)
		return true;	
	else
		return false;
}

function strlen(strsrc)
{
	return strsrc.length;
}

function strtrim(str)
{
	if(str)
		return str.replace(/(^\s*)|(\s*$)/g, "");
	else
		return "";
}

function linetrim(str)
{
	return str.replace(/(^[\r\n]*)|([\r\n]*$)/g, "");
}

///////////////////////////////////////////////////////////////////////////////////////////////
// vCalendar
function vCalendar()
{
	
}

vCalendar.prototype.PEOPLE = function (strOrg)
{
	var RSVP = "";
	var CN = "";
	var PARTSTAT = "";
	var ROLE = "";
	var MAILTO = "";
	var strSeg = strOrg.split(";");
	for(var i = 0; i < strSeg.length; i++)
	{
		if(strncasecmp(strSeg[i], "RSVP=", strlen("RSVP=")))
		{
			RSVP = strSeg[i].split("=")[1];
		}
		else if(strncasecmp(strSeg[i], "CN=", strlen("CN=")))
		{
			CN = strSeg[i].split("=")[1];
		}
		else if(strncasecmp(strSeg[i], "PARTSTAT=", strlen("PARTSTAT=")))
		{
			PARTSTAT = strSeg[i].split("=")[1];
		}
		else if(strncasecmp(strSeg[i], "ROLE=", strlen("ROLE=")))
		{
			ROLE = strSeg[i].split("=")[1].split(":")[0];
			MAILTO = strSeg[i].split("=")[1].split(":")[2];
		}
	}
	
	return [strtrim(RSVP), strtrim(CN), strtrim(PARTSTAT), strtrim(ROLE), strtrim(MAILTO)];
}

vCalendar.prototype.DT = function (strDT)
{
	var strSeg = strDT.split(":");
	var strTmp = strSeg[1] == null ? strSeg[0] : strSeg[1];
	return [strTmp.substr(0, 4), strTmp.substr(4, 2), strTmp.substr(6, 2), strTmp.substr(9, 2), strTmp.substr(11, 2), strTmp.substr(13, 2)];
}

vCalendar.prototype.TZID = function (strTZID)
{
	var strSeg = strTZID.split("=");
	var strTmp = strSeg[1] == null ? strSeg[0] : strSeg[1];
	return strTmp;
}

vCalendar.prototype.Parse = function (strcal)
{
	var VCALENDAR = (1 << 0);
	var VTIMEZONE = (1 << 1)
	var STANDARD = (1 << 2);
	var VEVENT = (1 << 3);
	
	var SEGMENT = 0;
	
	var calDetail = new JHashMap();
	var lArr = strcal.split("\n");
	
	for(var x = 0; x < lArr.length; x++)
	{
		var strTmp = linetrim(lArr[x]);
		var j = x + 1;
		while(j < lArr.length && (lArr[j].charAt(0) == ' ' || lArr[j].charAt(0) == '\t'))
		{
			strTmp += linetrim(lArr[j].substr(1));
			j++;
		}
		x = j - 1;
		
		if(strncasecmp(strTmp, "BEGIN:VCALENDAR", strlen("BEGIN:VCALENDAR")))
		{
			SEGMENT = SEGMENT | VCALENDAR;
		}
		else if(strncasecmp(strTmp, "END:VCALENDAR", strlen("END:VCALENDAR")))
		{
			SEGMENT = SEGMENT & (~VCALENDAR);
		}
		else if(strncasecmp(strTmp, "BEGIN:VTIMEZONE", strlen("BEGIN:VTIMEZONE")))
		{
			SEGMENT = SEGMENT | VTIMEZONE;
		}
		else if(strncasecmp(strTmp, "END:VTIMEZONE", strlen("END:VTIMEZONE")))
		{
			SEGMENT = SEGMENT & (~VTIMEZONE);
		}
		else if(strncasecmp(strTmp, "BEGIN:STANDARD", strlen("BEGIN:STANDARD")))
		{
			SEGMENT = SEGMENT | STANDARD;
		}
		else if(strncasecmp(strTmp, "END:STANDARD", strlen("END:STANDARD")))
		{
			SEGMENT = SEGMENT & (~STANDARD);
		}
		else if(strncasecmp(strTmp, "BEGIN:VEVENT", strlen("BEGIN:VEVENT")))
		{
			SEGMENT = SEGMENT | VEVENT;
		}
		else if(strncasecmp(strTmp, "END:VEVENT", strlen("END:VEVENT")))
		{
			SEGMENT = SEGMENT & (~VEVENT);
		}
		else
		{	
			if((SEGMENT & VCALENDAR) == VCALENDAR)
			{				
				if((SEGMENT & VTIMEZONE) == VTIMEZONE)
				{
					if(strncasecmp(strTmp, "TZID:", strlen("TZID:")))
					{
						calDetail.put("VTIMEZONE/TZID", strTmp.substr(strlen("TZID:")));
					}
					else if(strncasecmp(strTmp, "X-LIC-LOCATION:", strlen("X-LIC-LOCATION:")))
					{
						calDetail.put("VTIMEZONE/X-LIC-LOCATION", strTmp.substr(strlen("X-LIC-LOCATION:")));
					}
					if((SEGMENT & STANDARD) == STANDARD)
					{
						if(strncasecmp(strTmp, "TZOFFSETFROM:", strlen("TZOFFSETFROM:")))
						{
							calDetail.put("VTIMEZONE/STANDARD/TZOFFSETFROM", strTmp.substr(strlen("TZOFFSETFROM:")));	
						}
						else if(strncasecmp(strTmp, "TZOFFSETTO:", strlen("TZOFFSETTO:")))
						{
							calDetail.put("VTIMEZONE/STANDARD/TZOFFSETTO", strTmp.substr(strlen("TZOFFSETTO:")));
						}
						else if(strncasecmp(strTmp, "TZNAME:", strlen("TZNAME:")))
						{
							calDetail.put("VTIMEZONE/STANDARD/TZNAME", strTmp.substr(strlen("TZNAME:")));	
						}
						else if(strncasecmp(strTmp, "DTSTART:", strlen("DTSTART:")))
						{
							var dt = this.DT(strTmp.substr(strlen("DTSTART:")));
							calDetail.put("VTIMEZONE/STANDARD/DTSTART", dt[0] + "-" + dt[1] + "-" + dt[2] + " " + dt[3] + ":" + dt[4]);
						}
						else if(strncasecmp(strTmp, "DTEND:", strlen("DTEND:")))
						{
							var dt = this.DT(strTmp.substr(strlen("DTEND:")));
							calDetail.put("VTIMEZONE/STANDARD/DTEND", dt[0] + "-" + dt[1] + "-" + dt[2] + " " + dt[3] + ":" + dt[4]);
						}		
					}
				}
				else if((SEGMENT & VEVENT) == VEVENT)
				{
					if(strncasecmp(strTmp, "CREATED:", strlen("CREATED:")))
					{
						var dt = this.DT(strTmp.substr(strlen("CREATED:")));
						calDetail.put("VEVENT/CREATED", dt[0] + "-" + dt[1] + "-" + dt[2] + " " + dt[3] + ":" + dt[4]);	
					}
					else if(strncasecmp(strTmp, "LAST-MODIFIED:", strlen("LAST-MODIFIED:")))
					{
						var dt = this.DT(strTmp.substr(strlen("LAST-MODIFIED:")));
						calDetail.put("VEVENT/LAST-MODIFIED", dt[0] + "-" + dt[1] + "-" + dt[2] + " " + dt[3] + ":" + dt[4])	
					}
					else if(strncasecmp(strTmp, "DTSTAMP:", strlen("DTSTAMP:")))
					{
						var dt = this.DT(strTmp.substr(strlen("DTSTAMP:")));
						calDetail.put("VEVENT/DTSTAMP", dt[0] + "-" + dt[1] + "-" + dt[2] + " " + dt[3] + ":" + dt[4]);	
					}
					else if(strncasecmp(strTmp, "UID:", strlen("UID:")))
					{
						calDetail.put("VEVENT/UID", strTmp.substr(strlen("UID:")));	
					}
					else if(strncasecmp(strTmp, "SUMMARY:", strlen("SUMMARY:")))
					{
						strTmp = strTmp.substr(strlen("SUMMARY:")).replace(/(\\n|\\N)/g, "\n");
						strTmp = strTmp.replace(/\\\\/g, "\\");
						strTmp = strTmp.replace(/\\\"/g, "\"");
						strTmp = strTmp.replace(/\\t/g, "\t");
						strTmp = strTmp.replace(/\\;/g, ";");
						strTmp = strTmp.replace(/\\,/g, ",");
						
						calDetail.put("VEVENT/SUMMARY", strTmp);	
					}
					else if(strncasecmp(strTmp, "LOCATION:", strlen("LOCATION:")))
					{
						strTmp = strTmp.substr(strlen("LOCATION:")).replace(/(\\n|\\N)/g, "\n");
						strTmp = strTmp.replace(/\\\\/g, "\\");
						strTmp = strTmp.replace(/\\\"/g, "\"");
						strTmp = strTmp.replace(/\\t/g, "\t");
						strTmp = strTmp.replace(/\\;/g, ";");
						strTmp = strTmp.replace(/\\,/g, ",");
						
						calDetail.put("VEVENT/LOCATION", strTmp);	
					}
					else if(strncasecmp(strTmp, "DESCRIPTION:", strlen("DESCRIPTION:")))
					{
						strTmp = strTmp.substr(strlen("DESCRIPTION:")).replace(/(\\n|\\N)/g, "\n");
						strTmp = strTmp.replace(/\\\\/g, "\\");
						strTmp = strTmp.replace(/\\\"/g, "\"");
						strTmp = strTmp.replace(/\\t/g, "\t");
						strTmp = strTmp.replace(/\\;/g, ";");
						strTmp = strTmp.replace(/\\,/g, ",");
						
						calDetail.put("VEVENT/DESCRIPTION", strTmp);
					}
					else if(strncasecmp(strTmp, "SEQUENCE:", strlen("SEQUENCE:")))
					{
						calDetail.put("VEVENT/SEQUENCE", strTmp.substr(strlen("SEQUENCE:")));
					}
					else
					{
						if(strncasecmp(strTmp, "ORGANIZER;", strlen("ORGANIZER;")))
						{
							var people = this.PEOPLE(strTmp);
							calDetail.put("VEVENT/ORGANIZER", calDetail.get("VEVENT/ORGANIZER") == null ? (people[4] + "|" + people[2]) : (calDetail.get("VEVENT/ORGANIZER") + ";" + people[4] + "|" + people[2]));
						}
						else if(strncasecmp(strTmp, "ATTENDEE;", strlen("ATTENDEE;")))
						{
							var people = this.PEOPLE(strTmp);
							calDetail.put("VEVENT/ATTENDEE", calDetail.get("VEVENT/ATTENDEE") == null ? (people[4] + "|" + people[2]) : (calDetail.get("VEVENT/ATTENDEE") + ";" + people[4] + "|" + people[2]))
						}
						else if(strncasecmp(strTmp, "DTSTART;", strlen("DTSTART;")))
						{
							var strDTSTART = strTmp.split(";")[1].split(":");
							var tzid = this.TZID(strDTSTART[0]);
							var dt = this.DT(strDTSTART[1]);
							calDetail.put("VEVENT/DTSTART/TZID", tzid);
							calDetail.put("VEVENT/DTSTART/DT", dt[0] + "-" + dt[1] + "-" + dt[2] + " " + dt[3] + ":" + dt[4]);
						}
						else if(strncasecmp(strTmp, "DTEND;", strlen("DTEND;")))
						{
							var strDTEND = strTmp.split(";")[1].split(":");
							var tzid = this.TZID(strDTEND[0]);
							var dt = this.DT(strDTEND[1]);
							calDetail.put("VEVENT/DTEND/TZID", tzid);
							calDetail.put("VEVENT/DTEND/DT", dt[0] + "-" + dt[1] + "-" + dt[2] + " " + dt[3] + ":" + dt[4]);
						}
					}
				}
				else
				{
					if(strncasecmp(strTmp, "VERSION:", strlen("VERSION:")))
					{
						calDetail.put("VERSION", strTmp.substr(strlen("VERSION:")));	
					}
					else if(strncasecmp(strTmp, "METHOD:", strlen("METHOD:")))
					{
						calDetail.put("METHOD", strTmp.substr(strlen("METHOD:")));
					}
				}
			}
		}
	}
	
	return calDetail;
}

File: src/calendar.cpp
/*
        Copyright (c) openheap, uplusware
        uplusware@gmail.com
*/
#include "calendar.h"
#include "base.h"

Calendar::Calendar(const char* owner) {
  m_owner = owner;
  m_details.clear();
  m_phase = eNull;
  calbuf = "";
}

Calendar::~Calendar() {}

void Calendar::_parse_() {
  if (calbuf != "") {
    _calendar.text += calbuf;

    string strTmp;

    _strdelete_(calbuf, "\r\n ");
    _strdelete_(calbuf, "\r\n \t");
    _strdelete_(calbuf, "\n ");
    _strdelete_(calbuf, "\n\t");

    if (m_phase == eNull && strncasecmp(calbuf.c_str(), "BEGIN:VCALENDAR",
                                        sizeof("BEGIN:VCALENDAR") - 1) == 0) {
      _calendar.text = calbuf;
      _calendar.method = "";
      _calendar.prod = "";
      _calendar.version = "";
      _calendar.events.clear();
      _calendar.time_zones.clear();

      m_phase = eCal;
    } else if (m_phase == eCal &&
               strncasecmp(calbuf.c_str(), "PRODID:", sizeof("PRODID:") - 1) ==
                   0) {
      fnfy_strcut(calbuf.c_str(), "PRODID:", " \t\r\n", " \t\r\n",
                  _calendar.prod);
    } else if (m_phase == eCal && strncasecmp(calbuf.c_str(), "VERSION:",
                                              sizeof("VERSION:") - 1) == 0) {
      fnfy_strcut(calbuf.c_str(), "VERSION:", " \t\r\n", " \t\r\n",
                  _calendar.version);
    } else if (m_phase == eCal &&
               strncasecmp(calbuf.c_str(), "METHOD:", sizeof("METHOD:") - 1) ==
                   0) {
      fnfy_strcut(calbuf.c_str(), "METHOD:", " \t\r\n", " \t\r\n",
                  _calendar.method);
    } else if (m_phase == eCal &&
               strncasecmp(calbuf.c_str(), "BEGIN:VTIMEZONE",
                           sizeof("BEGIN:VTIMEZONE") - 1) == 0) {
      m_phase = eTimeZone;

      _time_zone.tid = "";
      _time_zone.xlic_location = "";
      _time_zone.standards.clear();
    } else if (m_phase == eTimeZone &&
               strncasecmp(calbuf.c_str(), "TZID:", sizeof("TZID:") - 1) == 0) {
      fnfy_strcut(calbuf.c_str(), "TZID:", " \t\r\n", " \t\r\n",
                  _time_zone.tid);
    } else if (m_phase == eTimeZone &&
               strncasecmp(calbuf.c_str(), "X-LIC-LOCATION:",
                           sizeof("X-LIC-LOCATION:") - 1) == 0) {
      fnfy_strcut(calbuf.c_str(), "X-LIC-LOCATION:", " \t\r\n", " \t\r\n",
                  _time_zone.xlic_location);
    }

    else if (m_phase == eTimeZone &&
             strncasecmp(calbuf.c_str(), "BEGIN:STANDARD",
                         sizeof("BEGIN:STANDARD") - 1) == 0) {
      m_phase == eStandard;

      _standard.offsetfr = 0;
      _standard.offsetto = 0;
      _standard.tzname = "";
      _standard.dtstart = 0;
    } else if (m_phase == eStandard &&
               strncasecmp(calbuf.c_str(),
                           "TZOFFSETFROM:", sizeof("TZOFFSETFROM:") - 1) == 0) {
      fnfy_strcut(calbuf.c_str(), "TZOFFSETFROM:", " \t\r\n", " \t\r\n",
                  strTmp);
      sscanf(strTmp.c_str(), "%d", _standard.offsetfr);
    } else if (m_phase == eStandard &&
               strncasecmp(calbuf.c_str(),
                           "TZOFFSETTO:", sizeof("TZOFFSETTO:") - 1) == 0) {
      fnfy_strcut(calbuf.c_str(), "TZOFFSETFROM:", " \t\r\n", " \t\r\n",
                  strTmp);
      sscanf(strTmp.c_str(), "%d", _standard.offsetto);
    } else if (m_phase == eStandard &&
               strncasecmp(calbuf.c_str(), "TZNAME:", sizeof("TZNAME:") - 1) ==
                   0) {
      fnfy_strcut(calbuf.c_str(), "TZNAME:", " \t\r\n", " \t\r\n",
                  _standard.tzname);
    } else if (m_phase == eStandard &&
               strncasecmp(calbuf.c_str(),
                           "DTSTART:", sizeof("DTSTART:") - 1) == 0) {
      fnfy_strcut(calbuf.c_str(), "DTSTART:", " \t\r\n", " \t\r\n", strTmp);

      struct tm dtTmp;

      sscanf(strTmp.c_str(), "%04d%02d%02dT%02d%02d%02dZ", &dtTmp.tm_year,
             &dtTmp.tm_mon, &dtTmp.tm_mday, &dtTmp.tm_hour, &dtTmp.tm_min,
             &dtTmp.tm_sec);
      dtTmp.tm_year -= 1900;
      dtTmp.tm_mon -= 1;

      _standard.dtstart = mktime(&dtTmp);

    } else if (m_phase == eStandard &&
               strncasecmp(calbuf.c_str(), "END:STANDARD",
                           sizeof("END:STANDARD") - 1) == 0) {
      _time_zone.standards.push_back(_standard);
      m_phase == eTimeZone;
    } else if (m_phase == eTimeZone &&
               strncasecmp(calbuf.c_str(), "END:VTIMEZONE",
                           sizeof("END:VTIMEZONE") - 1) == 0) {
      m_phase = eCal;

      _calendar.time_zones.push_back(_time_zone);
    } else if (m_phase == eCal &&
               strncasecmp(calbuf.c_str(), "BEGIN:VEVENT",
                           sizeof("BEGIN:VEVENT") - 1) == 0) {
      m_phase = eEvent;

      _event.created = 0;
      _event.modified = 0;
      _event.stamp = 0;

      _event.uid = "";
      _event.summary = "";

      _event.orgnizer.rsvp = true;
      _event.orgnizer.name = "";
      _event.orgnizer.role = "";
      _event.orgnizer.mailto = "";

      _event.attendees.clear();

      _event.start.tid = "";
      _event.start.dt = 0;

      _event.end.tid = "";
      _event.end.dt = 0;

      _event.location = "";
      _event.description = "";
    } else if (m_phase == eEvent && strncasecmp(calbuf.c_str(), "CREATED:",
                                                sizeof("CREATED:") - 1) == 0) {
      fnfy_strcut(calbuf.c_str(), "CREATED:", " \t\r\n", " \t\r\n", strTmp);

      struct tm dtTmp;

      sscanf(strTmp.c_str(), "%04d%02d%02dT%02d%02d%02dZ", &dtTmp.tm_year,
             &dtTmp.tm_mon, &dtTmp.tm_mday, &dtTmp.tm_hour, &dtTmp.tm_min,
             &dtTmp.tm_sec);
      dtTmp.tm_year -= 1900;
      dtTmp.tm_mon -= 1;

      _event.created = mktime(&dtTmp);
    } else if (m_phase == eEvent &&
               strncasecmp(calbuf.c_str(), "LAST-MODIFIED:",
                           sizeof("LAST-MODIFIED:") - 1) == 0) {
      fnfy_strcut(calbuf.c_str(), "LAST-MODIFIED:", " \t\r\n", " \t\r\n",
                  strTmp);

      struct tm dtTmp;

      sscanf(strTmp.c_str(), "%04d%02d%02dT%02d%02d%02dZ", &dtTmp.tm_year,
             &dtTmp.tm_mon, &dtTmp.tm_mday, &dtTmp.tm_hour, &dtTmp.tm_min,
             &dtTmp.tm_sec);
      dtTmp.tm_year -= 1900;
      dtTmp.tm_mon -= 1;

      _event.modified = mktime(&dtTmp);

    } else if (m_phase == eEvent && strncasecmp(calbuf.c_str(), "DTSTAMP:",
                                                sizeof("DTSTAMP:") - 1) == 0) {
      fnfy_strcut(calbuf.c_str(), "DTSTAMP:", " \t\r\n", " \t\r\n", strTmp);

      struct tm dtTmp;

      sscanf(strTmp.c_str(), "%04d%02d%02dT%02d%02d%02dZ", &dtTmp.tm_year,
             &dtTmp.tm_mon, &dtTmp.tm_mday, &dtTmp.tm_hour, &dtTmp.tm_min,
             &dtTmp.tm_sec);
      dtTmp.tm_year -= 1900;
      dtTmp.tm_mon -= 1;

      _event.stamp = mktime(&dtTmp);

    } else if (m_phase == eEvent &&
               strncasecmp(calbuf.c_str(), "UID:", sizeof("UID:") - 1) == 0) {
      fnfy_strcut(calbuf.c_str(), "UID:", " \t\r\n", " \t\r\n", _event.uid);
    } else if (m_phase == eEvent && strncasecmp(calbuf.c_str(), "SUMMARY:",
                                                sizeof("SUMMARY:") - 1) == 0) {
      fnfy_strcut(calbuf.c_str(), "UID:", " \t\r\n", " \t\r\n", _event.summary);
    } else if (m_phase == eEvent &&
               strncasecmp(calbuf.c_str(), "ORGANIZER;",
                           sizeof("ORGANIZER;") - 1) == 0) {
      fnfy_strcut(calbuf.c_str(), "RSVP=", " \t\r\n", "; \t\r\n",
                  _event.orgnizer.rsvp);

      fnfy_strcut(calbuf.c_str(), "CN=", " \t\r\n", "; \t\r\n",
                  _event.orgnizer.name);

      fnfy_strcut(calbuf.c_str(), "PARTSTAT=", " \t\r\n", "; \t\r\n",
                  _event.orgnizer.partstat);

      fnfy_strcut(calbuf.c_str(), "ROLE=", " \t\r\n", ":; \t\r\n",
                  _event.orgnizer.role);

      fnfy_strcut(calbuf.c_str(), "mailto:", " \t\r\n", "; \t\r\n",
                  _event.orgnizer.mailto);

    } else if (m_phase == eEvent && strncasecmp(calbuf.c_str(), "ATTENDEE;",
                                                sizeof("ATTENDEE;") - 1) == 0) {
      fnfy_strcut(calbuf.c_str(), "RSVP=", " \t\r\n", "; \t\r\n",
                  _attendee.rsvp);

      fnfy_strcut(calbuf.c_str(), "CN=", " \t\r\n", "; \t\r\n", _attendee.name);

      fnfy_strcut(calbuf.c_str(), "PARTSTAT=", " \t\r\n", "; \t\r\n",
                  _attendee.partstat);

      fnfy_strcut(calbuf.c_str(), "ROLE=", " \t\r\n", ":; \t\r\n",
                  _attendee.role);

      fnfy_strcut(calbuf.c_str(), "mailto:", " \t\r\n", "; \t\r\n",
                  _attendee.mailto);

      _event.attendees.push_back(_attendee);
    } else if (m_phase == eEvent && strncasecmp(calbuf.c_str(), "DTSTART;",
                                                sizeof("DTSTART;") - 1) == 0) {
      fnfy_strcut(calbuf.c_str(), ":", " \t\r\n", " \t\r\n", strTmp);

      struct tm dtTmp;

      sscanf(strTmp.c_str(), "%04d%02d%02dT%02d%02d%02d", &dtTmp.tm_year,
             &dtTmp.tm_mon, &dtTmp.tm_mday, &dtTmp.tm_hour, &dtTmp.tm_min,
             &dtTmp.tm_sec);
      dtTmp.tm_year -= 1900;
      dtTmp.tm_mon -= 1;

      _event.start.dt = mktime(&dtTmp);

      fnfy_strcut(calbuf.c_str(), "TZID=", " \t\r\n", ": \t\r\n", strTmp);
      _event.start.tid = strTmp;
    } else if (m_phase == eEvent && strncasecmp(calbuf.c_str(), "DTEND;",
                                                sizeof("DTEND;") - 1) == 0) {
      fnfy_strcut(calbuf.c_str(), ":", " \t\r\n", " \t\r\n", strTmp);

      struct tm dtTmp;

      sscanf(strTmp.c_str(), "%04d%02d%02dT%02d%02d%02d", &dtTmp.tm_year,
             &dtTmp.tm_mon, &dtTmp.tm_mday, &dtTmp.tm_hour, &dtTmp.tm_min,
             &dtTmp.tm_sec);
      dtTmp.tm_year -= 1900;
      dtTmp.tm_mon -= 1;

      _event.end.dt = mktime(&dtTmp);

      fnfy_strcut(calbuf.c_str(), "TZID=", " \t\r\n", ": \t\r\n", strTmp);
      _event.end.tid = strTmp;
    } else if (m_phase == eEvent && strncasecmp(calbuf.c_str(), "LOCATION:",
                                                sizeof("LOCATION:") - 1) == 0) {
      fnfy_strcut(calbuf.c_str(), "LOCATION:", " \t\r\n", " \t\r\n",
                  _event.location);
    } else if (m_phase == eEvent &&
               strncasecmp(calbuf.c_str(),
                           "DESCRIPTION:", sizeof("DESCRIPTION:") - 1) == 0) {
      fnfy_strcut(calbuf.c_str(), "DESCRIPTION:", " \t\r\n", " \t\r\n",
                  _event.description);
    } else if (m_phase == eEvent &&
               strncasecmp(calbuf.c_str(), "TRANSP:", sizeof("TRANSP:") - 1) ==
                   0) {
      fnfy_strcut(calbuf.c_str(), "TRANSP:", " \t\r\n", " \t\r\n",
                  _event.transp);
    } else if (m_phase == eEvent && strncasecmp(calbuf.c_str(), "SEQUENCE:",
                                                sizeof("SEQUENCE:") - 1) == 0) {
      fnfy_strcut(calbuf.c_str(), "SEQUENCE:", " \t\r\n", " \t\r\n", strTmp);
      sscanf(strTmp.c_str(), "%lu", &_event.sequence);
    } else if (m_phase == eEvent &&
               strncasecmp(calbuf.c_str(), "END:VEVENT",
                           sizeof("END:VEVENT") - 1) == 0) {
      _calendar.events.push_back(_event);
      m_phase = eCal;
    } else if (m_phase == eCal &&
               strncasecmp(calbuf.c_str(), "END:VCALENDAR",
                           sizeof("END:VCALENDAR") - 1) == 0) {
      m_phase = eNull;
      m_details.push_back(_calendar);
    }
  }
}

void Calendar::parse(const char* text) {
  if (text[0] == ' ' || text[0] == '\t') {
    calbuf += text;
  } else {
    _parse_();

    calbuf = text;
  }
}

void Calendar::save() {
  flush();

  char calpath[1024];
  struct tm* tmTmp;
  for (int i = 0; i < m_details.size(); i++) {
    for (int x = 0; x < m_details[i].events.size(); x++) {
      tmTmp = gmtime(&m_details[i].events[x].start.dt);
      int a = sprintf(calpath, "%s/cal/%s", CMailBase::m_private_path.c_str(),
                      m_owner.c_str());
      mkdir(calpath, 0744);

      int b = sprintf(calpath + a, "/%d%02d%02d", tmTmp->tm_year + 1900,
                      tmTmp->tm_mon + 1, tmTmp->tm_mday);
      mkdir(calpath, 0744);

      sprintf(calpath + a + b, "/%08x_%08x_%016lx_%08x.cal", time(NULL),
              getpid(), pthread_self(), random());

      m_file = fopen(calpath, "wb+");
      if (m_file) {
        fwrite(m_details[i].text.c_str(), m_details[i].text.length(), 1,
               m_file);
        fclose(m_file);
      }
    }
  }
}

void Calendar::load(const char* path) {
  FILE* pfile;
  pfile = fopen(path, "rb");

  char buf[4096];
  if (pfile) {
    while (feof(pfile) == 0) {
      fgets(buf, 4096, pfile);
      parse(buf);
    }
    flush();
    fclose(pfile);
  }
}

void Calendar::flush() {
  _parse_();
  calbuf = "";
}
